projects/Mallard/src/services/logging.ts

```diff
@@ -0,0 +1,190 @@
+            iAP,
+            ...optionalFields,
+        }
+    }
+
+    async getQueuedLogs(): Promise<MallardLogFormat[] | [{}]> {
+        try {
+            const logString = await loggingQueueCache.get()
+            return JSON.parse(logString || '[{}]')
+        } catch (e) {
+            return [{}]
+        }
+    }
+
+    async saveQueuedLogs(log: MallardLogFormat[]): Promise<string | Error> {
+        try {
+            const logString = JSON.stringify(log)
+            await loggingQueueCache.set(logString)
+            return 'saved logs'
+        } catch (e) {
+            errorService.captureException(e)
+            throw new Error(e)
+        }
+    }
+
+    async queueLogs(log: MallardLogFormat[]) {
+        try {
+            const currentQueue = await this.getQueuedLogs()
+            const currentQueueString = JSON.stringify(currentQueue)
+            const parsedQueue = JSON.parse(currentQueueString || '[{}]')
+            const newQueue = [...parsedQueue, ...log]
+            const cleanLogs = newQueue.filter(
+                value => Object.keys(value).length !== 0,
+            )
+            return cleanLogs
+        } catch (e) {
+            errorService.captureException(e)
+            throw new Error(e)
+        }
+    }
+
+    async clearLogs() {
+        try {
+            return await loggingQueueCache.reset()
+        } catch (e) {
+            errorService.captureException(e)
+        }
+    }
+
+    async postLog(log: MallardLogFormat[]): Promise<Response | Error> {
+        try {
+            const response = await fetch(defaultSettings.logging, {
+                method: 'POST',
+                headers: {
+                    Accept: 'application/json',
+                    'Content-Type': 'application/json',
+                    apiKey: LOGGING_API_KEY,
+                },
+                body: JSON.stringify(log),
+            })
+            if (response.status !== 200) {
+                throw new Error(
+                    `Bad response from Logging Service - status: ${response.status}`,
+                )
+            }
+            return response
+        } catch (e) {
+            this.saveQueuedLogs(log)
+            throw new Error(e)
+        }
+    }
+
+    async log({ level, message, ...optionalFields }: LogParams) {
+        try {
+            const currentLog = await this.baseLog({
+                level,
+                message,
+                ...optionalFields,
+            })
+            const logsToPost = await this.queueLogs([currentLog])
+
+            const { isConnected } = await NetInfo.fetch()
+            // Not connected, save the log queue
+            if (!isConnected) {
+                return this.saveQueuedLogs(logsToPost)
+            }
+
+            const postLogToService = await this.postLog(logsToPost)
+            await this.clearLogs()
+            return postLogToService
+        } catch (e) {
+            errorService.captureException(e)
```

Comment:
Maybe we should `saveQueuedLogs` when there is network connection but it failed to post it?

Reply:
We do that here: https://github.com/guardian/editions/pull/1151/files/44a335a7f35f93ea14c99f20819484df7a669020#diff-9d3468b2e5f768fd7f1ebe3538bb8cd9R158